File: library/src/main/java/io/github/kobakei/materialfabspeeddial/FabSpeedDialMenu.java
package io.github.kobakei.materialfabspeeddial;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.v7.util.SortedList;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.SubMenu;

import java.util.ArrayList;
import java.util.List;

/**
 * Menu class of FAB speed dial.
 * This class implements {@link Menu} interface but some methods are marked as deprecated and not implemented.
 *
 * Created by keisuke on 2017/06/13.
 */
public class FabSpeedDialMenu implements Menu {

    @NonNull
    private final Context context;

    private SortedList<MenuItem> menuItems = new SortedList<>(MenuItem.class, new SortedList.Callback<MenuItem>() {
        @Override
        public int compare(MenuItem o1, MenuItem o2) {
            return o1.getOrder() - o2.getOrder();
        }

        @Override
        public void onChanged(int position, int count) {

        }

        @Override
        public boolean areContentsTheSame(MenuItem oldItem, MenuItem newItem) {
            return false;
        }

        @Override
        public boolean areItemsTheSame(MenuItem item1, MenuItem item2) {
            return false;
        }

        @Override
        public void onInserted(int position, int count) {

        }

        @Override
        public void onRemoved(int position, int count) {

        }

        @Override
        public void onMoved(int fromPosition, int toPosition) {

        }
    });

    public FabSpeedDialMenu(@NonNull Context context) {
        this.context = context;
    }

    @Override
    public MenuItem add(CharSequence title) {
        int itemId = menuItems.size() + 1;
        return add(0, itemId, itemId, title);
    }

    @Override
    public MenuItem add(@StringRes int titleRes) {
        int itemId = menuItems.size() + 1;
        return add(0, itemId, itemId, titleRes);
    }

    @Override
    public MenuItem add(int groupId, int itemId, int order, CharSequence title) {
        MenuItem menuItem = new FabSpeedDialMenuItem(context, itemId, groupId, order);
        menuItem.setTitle(title);
        menuItems.add(menuItem);
        return menuItem;
    }

    @Override
    public MenuItem add(int groupId, int itemId, int order, @StringRes int titleRes) {
        MenuItem menuItem = new FabSpeedDialMenuItem(context, itemId, groupId, order);
        menuItem.setTitle(titleRes);
        menuItems.add(menuItem);
        return menuItem;
    }

    @Deprecated
    @Override
    public SubMenu addSubMenu(CharSequence title) {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public SubMenu addSubMenu(@StringRes int titleRes) {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public SubMenu addSubMenu(int groupId, int itemId, int order, CharSequence title) {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public SubMenu addSubMenu(int groupId, int itemId, int order, @StringRes int titleRes) {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public int addIntentOptions(int groupId, int itemId, int order, ComponentName caller, Intent[] specifics, Intent intent, int flags, MenuItem[] outSpecificItems) {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public void removeItem(int id) {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public void removeGroup(int groupId) {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public void clear() {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public void setGroupCheckable(int group, boolean checkable, boolean exclusive) {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public void setGroupVisible(int group, boolean visible) {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public void setGroupEnabled(int group, boolean enabled) {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public boolean hasVisibleItems() {
        throw new RuntimeException("Not implemented");
    }

    @Deprecated
    @Override
    public MenuItem findItem(int id) {
        throw new RuntimeException("Not implemented");
    }

    @Override
    public int size() {
        return menuItems.size();
    }

    @Override
    public MenuItem getItem(int index) {
        return menuItems.get(index);
    }

    @Override
    public void close() {
        throw new RuntimeException("Not implemented");
    }

    @Override
    public boolean performShortcut(int keyCode, KeyEvent event, int flags) {
        throw new RuntimeException("Not implemented");
    }

    @Override
    public boolean isShortcutKey(int keyCode, KeyEvent event) {
        throw new RuntimeException("Not implemented");
    }

    @Override
    public boolean performIdentifierAction(int id, int flags) {
        throw new RuntimeException("Not implemented");
    }

    @Override
    public void setQwertyMode(boolean isQwerty) {
        throw new RuntimeException("Not implemented");
    }
}

File: library/src/main/java/io/github/kobakei/materialfabspeeddial/FabSpeedDialMenu.kt
package io.github.kobakei.materialfabspeeddial

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import androidx.annotation.StringRes
import androidx.recyclerview.widget.SortedList
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.SubMenu

/**
 * Menu class of FAB speed dial.
 * This class implements [Menu] interface but some methods are marked as deprecated and not implemented.
 *
 * Created by keisuke on 2017/06/13.
 */
class FabSpeedDialMenu(private val context: Context) : Menu {

    private val menuItems = SortedList(MenuItem::class.java, object : SortedList.Callback<MenuItem>() {
        override fun compare(o1: MenuItem, o2: MenuItem): Int = o1.order - o2.order

        override fun onChanged(position: Int, count: Int) {

        }

        override fun areContentsTheSame(oldItem: MenuItem, newItem: MenuItem): Boolean = false

        override fun areItemsTheSame(item1: MenuItem, item2: MenuItem): Boolean = false

        override fun onInserted(position: Int, count: Int) {

        }

        override fun onRemoved(position: Int, count: Int) {

        }

        override fun onMoved(fromPosition: Int, toPosition: Int) {

        }
    })

    override fun add(title: CharSequence): MenuItem {
        val itemId = menuItems.size() + 1
        return add(0, itemId, itemId, title)
    }

    override fun add(@StringRes titleRes: Int): MenuItem {
        val itemId = menuItems.size() + 1
        return add(0, itemId, itemId, titleRes)
    }

    override fun add(groupId: Int, itemId: Int, order: Int, title: CharSequence): MenuItem {
        val menuItem = FabSpeedDialMenuItem(context, itemId, groupId, order)
        menuItem.title = title
        menuItems.add(menuItem)
        return menuItem
    }

    override fun add(groupId: Int, itemId: Int, order: Int, @StringRes titleRes: Int): MenuItem {
        val menuItem = FabSpeedDialMenuItem(context, itemId, groupId, order)
        menuItem.setTitle(titleRes)
        menuItems.add(menuItem)
        return menuItem
    }

    @Deprecated("")
    override fun addSubMenu(title: CharSequence): SubMenu {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun addSubMenu(@StringRes titleRes: Int): SubMenu {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun addSubMenu(groupId: Int, itemId: Int, order: Int, title: CharSequence): SubMenu {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun addSubMenu(groupId: Int, itemId: Int, order: Int, @StringRes titleRes: Int): SubMenu {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun addIntentOptions(groupId: Int, itemId: Int, order: Int, caller: ComponentName, specifics: Array<Intent>, intent: Intent, flags: Int, outSpecificItems: Array<MenuItem>): Int {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun removeItem(id: Int) {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun removeGroup(groupId: Int) {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun clear() {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun setGroupCheckable(group: Int, checkable: Boolean, exclusive: Boolean) {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun setGroupVisible(group: Int, visible: Boolean) {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun setGroupEnabled(group: Int, enabled: Boolean) {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun hasVisibleItems(): Boolean {
        throw RuntimeException("Not implemented")
    }

    @Deprecated("")
    override fun findItem(id: Int): MenuItem {
        throw RuntimeException("Not implemented")
    }

    override fun size(): Int = menuItems.size()

    override fun getItem(index: Int): MenuItem = menuItems.get(index)

    override fun close() {
        throw RuntimeException("Not implemented")
    }

    override fun performShortcut(keyCode: Int, event: KeyEvent, flags: Int): Boolean {
        throw RuntimeException("Not implemented")
    }

    override fun isShortcutKey(keyCode: Int, event: KeyEvent): Boolean {
        throw RuntimeException("Not implemented")
    }

    override fun performIdentifierAction(id: Int, flags: Int): Boolean {
        throw RuntimeException("Not implemented")
    }

    override fun setQwertyMode(isQwerty: Boolean) {
        throw RuntimeException("Not implemented")
    }
}
